### ISSUE TITLE
Project migration to webpack 5

### ISSUE THREAD

**Issue by pomek** [ckeditor/ckeditor5#9799]
The list below contains (I hope) all tasks required for full migration to webpack 5. If something is missed, please, mention it in a comment. 

⚠️ Before starting the task, make sure that https://github.com/ckeditor/ckeditor5/issues/6770 is resolved.

### Tools

- Since all changes that will be made should also be described in the migration guide for integrators, it would be good to write key points for particular steps in the issue.
- [x] Bump the version in `ckeditor/ckeditor5` and `ckeditor/ckeditor5-dev`

    ```diff
    - "webpack": "^4.43.0",
    - "webpack-cli": "^3.3.11"
    + "webpack": "^5.37.0",
    + "webpack-cli": "^4.7.0"
    ```

    - Consider upgrading and adjusting the configuration for loaders and plugins to avoid checking the same things twice.
- [x] Verify whether `ckeditor5-build-*` packages are built properly.
- [x] Verify whether manual tests work.
    - Use `yarn run manual` for preparing the HTTP server,
    - Use `yarn run manual:verify` to check whether tests do not throw errors
- [x] Verify whether automated tests work.
    - Use `yarn run test ...` for executing the automated tests,
    - Check whether all options work (`--coverage` must be supported).
- [x] Verify whether DLLs work.
    - Use `yarn run dll:build`
    - It touches the internal repository as well.
- [x] Verify whether documentation works.
    - Use `yarn run docs` for building the documentation,
    - Make sure to adjust the `snippetadapter.js` module if needed,
    - Use `yarn run docs:serve` for starting the HTTP server,
    - Use `yarn run docs:verify` for running the crawler (the same like for manual tests)
    - Verify whether the content-styles extractor still works.
- [x] As we change the snippet adapter, may Umberto be changed as well?
- [x] Upgrade webapck in the package generator
- [x] Bump dependencies in Online Builder. Talk with Maciek or/and Paweł if you need help.
    - When preparing a build, use webpack 5.
    - Generated build inside the package should use webpack 5.
- [x] Verify whether Cloud Services uses webpack. Talk with Bartek.

### Documentation

- [x] Re-write or adjust guides in our documentation
    - Building from the source without frameworks (React, …)
    - Building from the source when implementing one of our integration (React, …)
    - Support for different languages (e.g. building the editor with polish translations)
    - Extracting CSS to a separate file
    - Module replacement via webpack (e.g. replacing an icon)
- [x] Create the migration guide for integrators.
    - Cover a case: "I cannot use webpack 5 yet."

### Nice to have

- [ ] Integrations
    - React (`ckeditor5-react`)
    - Vue (`ckeditor5-vue2` and `ckeditor5-vue`)
    - Angular (`ckeditor5-angular`)
- [ ] Does the `ckeditor5-inspector` require the upgrade?

## Loaders

1. `postcss-loader` for handling PostCSS syntax
    - [v4.0.0](https://github.com/webpack-contrib/postcss-loader/releases/tag/v4.0.0) changes the API in the webpack configuration and requires adding `css-loader`
    - [v5.0.0](https://github.com/webpack-contrib/postcss-loader/releases/tag/v5.0.0) requires webpack 5.
2. `style-laoder` for adding the `<script>` tag on a page
    - [v2.0.0](https://github.com/webpack-contrib/style-loader/releases/tag/v2.0.0) safe bump for us
3. `css-loader`
    - not used yet but `postcss-loader` (v4+) requires it
    - [v5.0.0](https://github.com/webpack-contrib/css-loader/releases/tag/v5.0.0)
4. `istanbul-instrumenter-loader` for checking the coverage. 
    - It does not work with DLLs imports.
    - We need to find an alternative.

## Plugins

- `ckeditor5-dev-webpack-plugin` already works in webpack 5.
- `mini-css-extract-plugin` used by the snippet adapter for extracting the CSS to a separate file.
- `karma-webpack` for automated tests ([see BC](https://github.com/ryanclark/karma-webpack/blob/master/CHANGELOG.md#500-alpha0-2018-12-13)).
- `terser-webpack-plugin` - used for minification the JS produced by webpack
    - Do we need it? Let's check what webpack 5 offers.
- `progress-bar-webpack-plugin` - used for displaying progress bar in the snippet adapter.

Most of the mentioned changes were covered in the PR – [https://github.com/ckeditor/ckeditor5-dev/pull/673/files](https://github.com/ckeditor/ckeditor5-dev/pull/673/files). Feel free to see whether it helps.

**Comment by pomek** [comment]
Regarding the documentation:

- Update guides to use the latest webpack (5)
- Migration guide: 
	- CKEditor 5 recommends using webpack 5
    - Upgrade `postcss-loader` configuration object
    - It requires adding the `css-loader` loader (show a code snippet)
    - Mention why we don't use the latest version of loaders (they drop support for webpack 4)

**Comment by pomek** [comment]
An issue about making CKEditor 5 dev utils compatible with webpack 5 – https://github.com/ckeditor/ckeditor5/issues/10735 – this one unblocks developers to prepare builds using webpack 5 (without waiting for decisions made by us).

**Comment by pomek** [comment]
Migration to webpack 5 finished. It will be a part of the upcoming release, which will happen next year (most probably, January).
